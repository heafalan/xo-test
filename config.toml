iso = 'Windows7Ultimate.iso'
network = 'Pool-wide network associated with eth0'
pool = 'lab3'
pvVm = 'xo-test-pv'
vmToMigrate = 'souad'
xoServerUrl = 'localhost:9000'
vmIdXoTest = '0342407f-9727-4f12-83c2-9d0c458916c2'
xoTestTemplateId = '1a33d04d-1a9c-e2af-ae42-010554eed513'

# powerOnMode has to be not empty
host1 = 'lab1'

[adminCredentials]
  email = 'admin@admin.net'
  password = 'admin'

[templates]
debian = 'Debian Wheezy 7.0 (64-bit)'
otherConfig = 'Other install media'
centOS = 'CentOS 7'

[remotes]
defaultRemote1 = '8565fc53-ea32-40ea-b380-dd1b8f3b5d44'
defaultRemote2 = '606ec797-dc74-4042-8cb5-828b66e684f8'

[srs]
defaultSr = 'a5954951-3dfa-42b8-803f-4bc270b22a0b'
srLocalStorage = 'ff33495d-68c4-0aa7-193a-229c261390b8'

[slaveServer]
host = '192.168.100.1'
username = 'root'
password = 'qwerty'
autoConnect = false

[[xenServers]]
host = '192.168.100.3'
username = 'root'
password = 'qwerty'

[[xenServers]]
host = '192.168.100.2'
username = 'root'
password = 'qwerty'
