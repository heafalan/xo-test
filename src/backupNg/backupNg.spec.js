/* eslint-env jest */

import { findKey, omit } from "lodash";
import { noSuchObject } from "xo-common/api-errors";

import config from "../_config";
import randomId from "../_randomId";
import xo from "../_xoConnection";

describe("backupNg", () => {
  let defaultBackupNg;

  beforeAll(() => {
    defaultBackupNg = {
      name: "default-backupNg",
      mode: "full",
      vms: {
        id: config.vmIdXoTest,
      },
      settings: {
        "": {
          reportWhen: "always",
        },
      },
    };
  });

  describe(".create() :", () => {
    it("creates a new backup job without schedules", async () => {
      const backupNg = await xo.createTempBackupNgJob(defaultBackupNg);
      expect(omit(backupNg, "id", "userId")).toMatchSnapshot();
      expect(backupNg.userId).toBe(xo._user.id);
    });

    it("creates a new backup job with schedules", async () => {
      const scheduleTempId = randomId();
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: {
            name: "scheduleTest",
            cron: "0 * * * * *",
          },
        },
        settings: {
          ...defaultBackupNg.settings,
          [scheduleTempId]: { snapshotRetention: 1 },
        },
      });

      const backupNgJob = await xo.call("backupNg.getJob", { id: jobId });

      expect(omit(backupNgJob, "id", "userId", "settings")).toMatchSnapshot();
      expect(backupNgJob.userId).toBe(xo._user.id);

      const settingKeys = Object.keys(backupNgJob.settings);
      expect(settingKeys.length).toBe(2);
      const scheduleId = settingKeys.find(key => key !== "");
      expect(backupNgJob.settings[scheduleId]).toEqual({
        snapshotRetention: 1,
      });

      const schedule = await xo.call("schedule.get", { id: scheduleId });
      expect(omit(schedule, "id", "jobId")).toMatchSnapshot();
      expect(schedule.jobId).toBe(jobId);
    });
  });

  describe(".delete() :", () => {
    it("deletes a backup job", async () => {
      const scheduleTempId = randomId();
      const { id: jobId } = await xo.call("backupNg.createJob", {
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: {
            name: "scheduleTest",
            cron: "0 * * * * *",
          },
        },
        settings: {
          ...defaultBackupNg.settings,
          [scheduleTempId]: { snapshotRetention: 1 },
        },
      });

      const backupNgJob = await xo.call("backupNg.getJob", { id: jobId });
      const scheduleId = findKey(backupNgJob.settings, {
        snapshotRetention: 1,
      });

      await xo.call("backupNg.deleteJob", { id: jobId });

      let isRejectedJobErrorValid = false;
      await xo.call("backupNg.getJob", { id: jobId }).catch(error => {
        isRejectedJobErrorValid = noSuchObject.is(error);
      });
      expect(isRejectedJobErrorValid).toBe(true);

      let isRejectedScheduleErrorValid = false;
      await xo.call("schedule.get", { id: scheduleId }).catch(error => {
        isRejectedScheduleErrorValid = noSuchObject.is(error);
      });
      expect(isRejectedScheduleErrorValid).toBe(true);
    });
  });

  describe(".runJob() :", () => {
    it("fails trying to run a backup job without schedule", async () => {
      const { id } = await xo.createTempBackupNgJob(defaultBackupNg);
      await expect(
        xo.call("backupNg.runJob", { id })
      ).rejects.toMatchSnapshot();
    });

    it("fails trying to run a backup job with no matching VMs", async () => {
      const scheduleTempId = randomId();
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: {
            name: "scheduleTest",
            cron: "0 * * * * *",
          },
        },
        settings: {
          [scheduleTempId]: { snapshotRetention: 1 },
        },
        vms: {
          id: config.vmIdXoTest,
          name: "test-vm-backupNg",
        },
      });

      const backupNgJob = await xo.call("backupNg.getJob", { id: jobId });
      const settingKeys = Object.keys(backupNgJob.settings);
      expect(settingKeys.length).toBe(1);

      await expect(
        xo.call("backupNg.runJob", { id: jobId, schedule: settingKeys[0] })
      ).rejects.toMatchSnapshot();
    });

    it("fails trying to run a backup job with non-existent vm", async () => {
      const scheduleTempId = randomId();
      const { id: jobId } = await xo.createTempBackupNgJob({
        ...defaultBackupNg,
        schedules: {
          [scheduleTempId]: {
            name: "scheduleTest",
            cron: "0 * * * * *",
          },
        },
        settings: {
          [scheduleTempId]: { snapshotRetention: 1 },
        },
        vms: {
          id: "non-existent-id",
        },
      });

      const backupNgJob = await xo.call("backupNg.getJob", { id: jobId });

      const settingKeys = Object.keys(backupNgJob.settings);
      expect(settingKeys.length).toBe(1);

      await xo.call("backupNg.runJob", { id: jobId, schedule: settingKeys[0] });
      const [log] = await xo.call("backupNg.getLogs", {
        scheduleId: settingKeys[0],
      });
      expect(log.warnings).toMatchSnapshot();
    });
  });

  test("execute three times a DR with 2 as retention", async () => {
    jest.setTimeout(3e4);
    const SR = await xo.getSrId();
    const vmId = await xo.createTempVm({
      name_label: "XO Test Temporary",
      name_description:
        "Creating a temporary vm to execute three times a DR & CR with 2 as retention",
      template: config.xoTestTemplateId,
      high_availability: "",
      VDIs: [
        {
          device: "0",
          size: 1,
          SR,
          type: "user",
        },
      ],
    });

    const scheduleTempId = randomId();
    const srDeleteFirstTrue = config.srLocalStorage1;
    const srDeleteFirstFalse = config.srLocalStorage2;
    const { id: jobId } = await xo.createTempBackupNgJob({
      ...defaultBackupNg,
      remotes: {
        id: {
          __or: [],
        },
      },
      schedules: {
        [scheduleTempId]: {
          name: "scheduleTest",
          cron: "0 * * * * *",
        },
      },
      settings: {
        [srDeleteFirstTrue]: {
          deleteFirst: true,
        },
        [srDeleteFirstFalse]: {
          deleteFirst: false,
        },
        [scheduleTempId]: {
          copyRetention: 2,
        },
        "": {
          reportWhen: "Never",
          fullInterval: 3,
        },
      },
      srs: {
        id: {
          __or: [srDeleteFirstTrue, srDeleteFirstFalse],
        },
      },
      vms: {
        id: vmId,
      },
    });

    const schedule = await xo.getSchedule({ jobId });
    expect(typeof schedule).toBe("object");
    await xo.call("backupNg.runJob", { id: jobId, schedule: schedule.id });
    await xo.call("backupNg.runJob", { id: jobId, schedule: schedule.id });
    await xo.call("backupNg.runJob", { id: jobId, schedule: schedule.id });

    const replicatedVms = [];
    for (const obj in xo.objects.all) {
      if (xo.objects.all[obj].other) {
        const {
          "xo:backup:vm": backupVm,
          "xo:backup:job": backupJob,
          "xo:backup:schedule": backupSchedule,
        } = xo.objects.all[obj].other;
        if (
          backupVm === vmId &&
          backupJob === jobId &&
          backupSchedule === schedule.id
        ) {
          replicatedVms.push(xo.objects.all[obj]);
        }
      }
    }

    const expected = [
      xo.objects.all[vmId].name_label,
      defaultBackupNg.name,
      expect.stringMatching(/[A-Z0-9]*/),
    ];

    for (let i = 0; i < replicatedVms.length; i++) {
      expect(replicatedVms[i].name_label.split(" - ")).toEqual(
        expect.arrayContaining(expected)
      );
      expect(replicatedVms[i].tags).toMatchSnapshot();
      expect(replicatedVms[i].high_availability).toBe("");
      await expect(
        xo.call("vm.start", { id: replicatedVms[i].id })
      ).rejects.toMatchSnapshot();
      await xo.call("vm.start", { id: replicatedVms[i].id, force: true });
    }
  });
});
