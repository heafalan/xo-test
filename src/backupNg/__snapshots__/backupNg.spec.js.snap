// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backupNg .createJob() : creates a new backup job with schedules 1`] = `
Object {
  "id": Any<String>,
  "mode": "full",
  "name": "default-backupNg",
  "settings": Any<Object>,
  "type": "backup",
  "userId": Any<String>,
  "vms": Object {
    "id": "0342407f-9727-4f12-83c2-9d0c458916c2",
  },
}
`;

exports[`backupNg .createJob() : creates a new backup job with schedules 2`] = `
Object {
  "cron": "0 * * * * *",
  "enabled": false,
  "id": Any<String>,
  "jobId": Any<String>,
  "name": "scheduleTest",
}
`;

exports[`backupNg .createJob() : creates a new backup job without schedules 1`] = `
Object {
  "id": Any<String>,
  "mode": "full",
  "name": "default-backupNg",
  "settings": Object {
    "": Object {
      "reportWhen": "never",
    },
  },
  "type": "backup",
  "userId": Any<String>,
  "vms": Object {
    "id": "0342407f-9727-4f12-83c2-9d0c458916c2",
  },
}
`;

exports[`backupNg .runJob() : fails trying to run a backup job with a VM without disks 1`] = `
Object {
  "data": Object {
    "mode": "full",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "jobName": "default-backupNg",
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "skipped",
}
`;

exports[`backupNg .runJob() : fails trying to run a backup job with a VM without disks 2`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "message": "no disks found",
    "name": "Error",
    "stack": Any<String>,
  },
  "start": Any<Number>,
  "status": "skipped",
}
`;

exports[`backupNg .runJob() : fails trying to run a backup job with no matching VMs 1`] = `[JsonRpcError: unknown error from the peer]`;

exports[`backupNg .runJob() : fails trying to run a backup job with non-existent vm 1`] = `
Array [
  Object {
    "data": Object {
      "vms": Array [
        "non-existent-id",
      ],
    },
    "message": "missingVms",
  },
]
`;

exports[`backupNg .runJob() : fails trying to run a backup job without schedule 1`] = `[JsonRpcError: invalid parameters]`;

exports[`backupNg .runJob() : fails trying to run backup job with remotes without exportRetention 1`] = `
Object {
  "data": Object {
    "mode": "full",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "jobName": "default-backupNg",
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "failure",
}
`;

exports[`backupNg .runJob() : fails trying to run backup job with remotes without exportRetention 2`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "result": Object {
    "message": "copy, export and snapshot retentions cannot both be 0",
    "name": "Error",
    "stack": Any<String>,
  },
  "start": Any<Number>,
  "status": "failure",
}
`;

exports[`backupNg execute three times a rolling snapshot with 2 as retention & revert to an old state 1`] = `
Object {
  "data": Object {
    "mode": "full",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "jobName": "default-backupNg",
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot with 2 as retention & revert to an old state 2`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": "snapshot",
  "result": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot with 2 as retention & revert to an old state 3`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot, a delta backup and a CR, revert to an old state & restore the backups 1`] = `
Object {
  "data": Object {
    "mode": "delta",
    "reportWhen": "never",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "jobId": Any<String>,
  "jobName": "default-backupNg",
  "message": "backup",
  "scheduleId": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot, a delta backup and a CR, revert to an old state & restore the backups 2`] = `
Object {
  "end": Any<Number>,
  "id": Any<String>,
  "message": "snapshot",
  "result": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot, a delta backup and a CR, revert to an old state & restore the backups 3`] = `
Object {
  "data": Object {
    "id": Any<String>,
    "type": "VM",
  },
  "end": Any<Number>,
  "id": Any<String>,
  "message": Any<String>,
  "start": Any<Number>,
  "status": "success",
}
`;

exports[`backupNg execute three times a rolling snapshot, a delta backup and a CR, revert to an old state & restore the backups 4`] = `
Array [
  "Continuous Replication",
]
`;

exports[`backupNg execute three times a rolling snapshot, a delta backup and a CR, revert to an old state & restore the backups 5`] = `[JsonRpcError: forbidden operation: Start]`;

exports[`backupNg execute three times a rolling snapshot, a delta backup and a CR, revert to an old state & restore the backups 6`] = `
Array [
  "Continuous Replication",
]
`;

exports[`backupNg execute three times a rolling snapshot, a delta backup and a CR, revert to an old state & restore the backups 7`] = `[JsonRpcError: forbidden operation: Start]`;
